File: typescript/src/ops/_internal/instrumentation/providers/anthropic.ts
/**
 * OpenTelemetry instrumentation for Anthropic SDK.
 *
 * Uses the @traceloop/instrumentation-anthropic package.
 */

import type { TracerProvider } from "@opentelemetry/api";

import {
  BaseInstrumentation,
  type ContentCaptureMode,
  type Instrumentor,
} from "@/ops/_internal/instrumentation/providers/base";

// Lazy import to avoid requiring the package if not used
let AnthropicInstrumentationClass: (new () => Instrumentor) | null = null;

/* v8 ignore start - requires optional dependency */
async function getAnthropicInstrumentationClass(): Promise<
  new () => Instrumentor
> {
  if (AnthropicInstrumentationClass) {
    return AnthropicInstrumentationClass;
  }

  try {
    // Dynamic import of optional dependency - uses type declaration from types.d.ts
    const { AnthropicInstrumentation } =
      await import("@traceloop/instrumentation-anthropic");
    AnthropicInstrumentationClass =
      AnthropicInstrumentation as new () => Instrumentor;
    return AnthropicInstrumentationClass;
  } catch {
    throw new Error(
      "Failed to import @traceloop/instrumentation-anthropic. " +
        "Please install it: npm install @traceloop/instrumentation-anthropic",
    );
  }
}
/* v8 ignore end */

// Singleton state
let anthropicInstance: AnthropicInstrumentationManager | null = null;
let anthropicInstrumentorClass: (new () => Instrumentor) | null = null;

/**
 * Manages OpenTelemetry instrumentation lifecycle for the Anthropic SDK.
 */
class AnthropicInstrumentationManager extends BaseInstrumentation<Instrumentor> {
  static instance(): AnthropicInstrumentationManager {
    if (!anthropicInstance) {
      anthropicInstance = new AnthropicInstrumentationManager();
    }
    return anthropicInstance;
  }

  protected createInstrumentor(): Instrumentor {
    if (!anthropicInstrumentorClass) {
      throw new Error(
        "Instrumentor class not loaded. Call instrumentAnthropic() instead of using the class directly.",
      );
    }
    return new anthropicInstrumentorClass();
  }

  protected configureContentCapture(captureContent: ContentCaptureMode): void {
    // Anthropic/Traceloop uses TRACELOOP_TRACE_CONTENT
    if (captureContent === "enabled") {
      this.setEnvVar("TRACELOOP_TRACE_CONTENT", "true");
    } else if (captureContent === "disabled") {
      this.setEnvVar("TRACELOOP_TRACE_CONTENT", "false");
    }
  }

  static resetForTesting(): void {
    if (anthropicInstance) {
      anthropicInstance.uninstrument();
    }
    anthropicInstance = null;
    anthropicInstrumentorClass = null;
  }
}

/**
 * Enable OpenTelemetry instrumentation for the Anthropic SDK.
 *
 * Uses the provided tracerProvider or the global OpenTelemetry tracer provider.
 *
 * @param options - Configuration options
 * @param options.tracerProvider - Optional tracer provider to use
 * @param options.captureContent - Controls whether to capture message content in spans
 *
 * @example
 * ```typescript
 * import { ops } from '@mirascope/core';
 *
 * // Configure first
 * ops.configure();
 *
 * // Enable Anthropic instrumentation
 * await ops.instrumentAnthropic();
 *
 * // With content capture enabled
 * await ops.instrumentAnthropic({ captureContent: "enabled" });
 * ```
 */
/* v8 ignore start - requires optional dependency */
export async function instrumentAnthropic(
  options: {
    tracerProvider?: TracerProvider;
    captureContent?: ContentCaptureMode;
  } = {},
): Promise<void> {
  anthropicInstrumentorClass = await getAnthropicInstrumentationClass();
  const manager = AnthropicInstrumentationManager.instance();
  manager.instrument(options);
}
/* v8 ignore end */

/**
 * Disable previously configured Anthropic instrumentation.
 */
export function uninstrumentAnthropic(): void {
  AnthropicInstrumentationManager.instance().uninstrument();
}

/**
 * Return whether Anthropic instrumentation is currently active.
 */
export function isAnthropicInstrumented(): boolean {
  return AnthropicInstrumentationManager.instance().isInstrumented;
}

/**
 * Reset Anthropic instrumentation for testing.
 */
export function resetAnthropicInstrumentationForTesting(): void {
  AnthropicInstrumentationManager.resetForTesting();
}

File: python/mirascope/ops/_internal/instrumentation/providers/anthropic.py
"""OpenTelemetry instrumentation for Anthropic SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING

from opentelemetry.instrumentation.anthropic import AnthropicInstrumentor

from .base import BaseInstrumentation, ContentCaptureMode

if TYPE_CHECKING:
    from opentelemetry.trace import TracerProvider


class AnthropicInstrumentation(BaseInstrumentation[AnthropicInstrumentor]):
    """Manages OpenTelemetry instrumentation lifecycle for the Anthropic SDK."""

    def _create_instrumentor(self) -> AnthropicInstrumentor:
        """Create a new Anthropic instrumentor instance."""
        return AnthropicInstrumentor()

    def _configure_capture_content(self, capture_content: ContentCaptureMode) -> None:
        """Configure environment variables for Anthropic content capture."""
        if capture_content == "enabled":
            self._set_env_var("TRACELOOP_TRACE_CONTENT", "true")
        elif capture_content == "disabled":
            self._set_env_var("TRACELOOP_TRACE_CONTENT", "false")


def instrument_anthropic(
    *,
    tracer_provider: TracerProvider | None = None,
    capture_content: ContentCaptureMode = "default",
) -> None:
    """Enable OpenTelemetry instrumentation for the Anthropic Python SDK.

    Uses the provided tracer_provider or the global OpenTelemetry tracer provider.

    Args:
        tracer_provider: Optional tracer provider to use. If not provided,
            uses the global OpenTelemetry tracer provider.
        capture_content: Controls whether to capture message content in spans.
            - "enabled": Capture message content
            - "disabled": Do not capture message content
            - "default": Use the library's default behavior

    Example:

        Enable instrumentation with Mirascope Cloud:
        ```python
        from mirascope import ops

        ops.configure()
        ops.instrument_anthropic()
        ```

        Enable instrumentation with content capture:
        ```python
        from mirascope import ops

        ops.configure()
        ops.instrument_anthropic(capture_content="enabled")
        ```
    """
    AnthropicInstrumentation().instrument(
        tracer_provider=tracer_provider,
        capture_content=capture_content,
    )


def uninstrument_anthropic() -> None:
    """Disable previously configured Anthropic instrumentation."""
    AnthropicInstrumentation().uninstrument()


def is_anthropic_instrumented() -> bool:
    """Return whether Anthropic instrumentation is currently active."""
    return AnthropicInstrumentation().is_instrumented
